File: ios/RNKarteTrackerModule.m
//
//  RNKarteTrackerModule.m
//  react-native-karte-tracker
//
//  Created by tomoki.koga on 2018/12/03.
//

#import "RNKarteTrackerModule.h"

#import <KarteTracker/KarteTracker.h>

@interface KarteVariable ()
@property (nonatomic, copy, nullable, readwrite) NSString *campaignId;
@property (nonatomic, copy, nullable, readwrite) NSString *shortenId;
@end

@implementation RNKarteTrackerModule

RCT_EXPORT_MODULE()

- (dispatch_queue_t)methodQueue
{
    return dispatch_get_main_queue();
}

#pragma mark - KarteTracker
RCT_EXPORT_METHOD(getAppKey:(RCTResponseSenderBlock)callback)
{
    NSString *appKey = [[KarteTracker sharedTracker] appKey];
    callback(@[appKey]);
}

RCT_EXPORT_METHOD(getVisitorId:(RCTResponseSenderBlock)callback)
{
    NSString *visitorId = [[KarteTracker sharedTracker] visitorId];
    callback(@[visitorId]);
}

RCT_EXPORT_METHOD(track:(NSString *)eventName values:(NSDictionary *)values)
{
    [[KarteTracker sharedTracker] track:eventName values:values];
}

RCT_EXPORT_METHOD(identify:(NSDictionary *)values)
{
    [[KarteTracker sharedTracker] identify:values];
}

RCT_EXPORT_METHOD(view:(NSString *)viewName title:(NSString *)title values:(NSDictionary *)values)
{
    [[KarteTracker sharedTracker] view:viewName title:title values:values];
}

RCT_EXPORT_METHOD(registerFCMToken:(NSString *)fcmToken)
{
    [[KarteTracker sharedTracker] registerFCMToken:fcmToken];
}

RCT_EXPORT_METHOD(optIn)
{
    [[KarteTracker sharedTracker] optIn];
}

RCT_EXPORT_METHOD(optOut)
{
    [[KarteTracker sharedTracker] optOut];
}

RCT_EXPORT_METHOD(renewVisitorId)
{
    [[KarteTracker sharedTracker] renewVisitorId];
}

#pragma mark - KarteTrackerJsUtil
RCT_EXPORT_METHOD(appendUserSyncQueryParameter:(NSString *)url callback:(RCTResponseSenderBlock)callback)
{
    NSString *appKey = [[KarteTracker sharedTracker] appKey];
    NSString *newURL = [KarteTrackerJsUtil stringByAppendingUserSyncQueryParameter:appKey withURLString:url];
    callback(@[newURL]);
}

#pragma mark - KarteInAppMessagingManager
RCT_EXPORT_METHOD(isPresenting:(RCTResponseSenderBlock)callback)
{
    BOOL isPresenting = [[KarteInAppMessagingManager sharedManager] isPresenting];
    callback(@[@(isPresenting)]);
}

RCT_EXPORT_METHOD(dismiss)
{
    [[KarteInAppMessagingManager sharedManager] dismiss];
}

RCT_EXPORT_METHOD(suppress)
{
    [[KarteInAppMessagingManager sharedManager] suppress];
}

RCT_EXPORT_METHOD(unsuppress)
{
    [[KarteInAppMessagingManager sharedManager] unsuppress];
}

#pragma mark - KarteVariables
RCT_EXPORT_METHOD(fetch:(RCTResponseSenderBlock)callback)
{
    [KarteVariables fetchWithCompletionBlock:^(BOOL isSuccessful) {
        callback(@[@(isSuccessful)]);
    }];
}

RCT_EXPORT_METHOD(variable:(NSString *)key callback:(RCTResponseSenderBlock)callback)
{
    KarteVariable *variable = [KarteVariables variableForKey:key];
    
    NSMutableDictionary *dict = [NSMutableDictionary dictionary];
    [dict setObject:key forKey:@"key"];
    
    if (variable.campaignId) {
        [dict setObject:variable.campaignId forKey:@"campaign_id"];
    }
    if (variable.shortenId) {
        [dict setObject:variable.shortenId forKey:@"shorten_id"];
    }
    if (variable.string) {
        [dict setObject:variable.string forKey:@"string"];
    }
    
    callback(@[dict]);
}

RCT_EXPORT_METHOD(trackVariables:(NSArray *)vals eventName:(NSString *)eventName values:(NSDictionary *)values)
{
    NSMutableArray *variables = [NSMutableArray array];
    for (NSDictionary *val in vals) {
        KarteVariable *variable = [[KarteVariable alloc] init];
        variable.campaignId = val[@"campaign_id"];
        variable.shortenId = val[@"shorten_id"];
        [variables addObject:variable];
    }
    [KarteVariables trackWithVariables:variables withEventName:eventName withValues:values];
}

#pragma mark - KarteVariable
RCT_EXPORT_METHOD(stringForKey:(NSString *)key defaultValue:(NSString *)defaultValue callback:(RCTResponseSenderBlock)callback)
{
    KarteVariable *variable = [KarteVariables variableForKey:key];
    NSString *value = [variable stringWithDefaultValue:defaultValue];
    
    NSMutableArray *returns = [NSMutableArray array];
    if (value) {
        [returns addObject:value];
    }
    
    callback(returns);
}

RCT_EXPORT_METHOD(integerForKey:(NSString *)key defaultValue:(NSInteger)defaultValue callback:(RCTResponseSenderBlock)callback)
{
    KarteVariable *variable = [KarteVariables variableForKey:key];
    NSInteger value = [variable integerWithDefaultValue:defaultValue];
    callback(@[@(value)]);
}

RCT_EXPORT_METHOD(doubleForKey:(NSString *)key defaultValue:(double)defaultValue callback:(RCTResponseSenderBlock)callback)
{
    KarteVariable *variable = [KarteVariables variableForKey:key];
    double value = [variable doubleWithDefaultValue:defaultValue];
    callback(@[@(value)]);
}

RCT_EXPORT_METHOD(boolForKey:(NSString *)key defaultValue:(BOOL)defaultValue callback:(RCTResponseSenderBlock)callback)
{
    KarteVariable *variable = [KarteVariables variableForKey:key];
    bool value = [variable boolWithDefaultValue:defaultValue];
    callback(@[@(value)]);
}

RCT_EXPORT_METHOD(arrayForKey:(NSString *)key defaultValue:(NSArray *)defaultValue callback:(RCTResponseSenderBlock)callback)
{
    KarteVariable *variable = [KarteVariables variableForKey:key];
    NSArray *value = [variable arrayWithDefaultValue:defaultValue];
    
    NSMutableArray *returns = [NSMutableArray array];
    if (value) {
        [returns addObject:value];
    }
    
    callback(returns);
}

RCT_EXPORT_METHOD(objectForKey:(NSString *)key defaultValue:(NSDictionary *)defaultValue callback:(RCTResponseSenderBlock)callback)
{
    KarteVariable *variable = [KarteVariables variableForKey:key];
    NSDictionary *value = [variable dictionaryWithDefaultValue:defaultValue];
    
    NSMutableArray *returns = [NSMutableArray array];
    if (value) {
        [returns addObject:value];
    }
    
    callback(returns);
}

@end

File: android/src/main/java/io/karte/reactnative/RNKarteTrackerModule.java

package io.karte.reactnative;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Callback;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableMap;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

import io.karte.android.tracker.Tracker;
import io.karte.android.tracker.TrackerJsUtil;
import io.karte.android.tracker.Variable;
import io.karte.android.tracker.Variables;
import io.karte.android.tracker.inappmessaging.InAppMessagingManager;

public class RNKarteTrackerModule extends ReactContextBaseJavaModule {

  private final ReactApplicationContext reactContext;

  public RNKarteTrackerModule(ReactApplicationContext reactContext) {
    super(reactContext);
    this.reactContext = reactContext;
  }

  @Override
  public String getName() {
    return "RNKarteTrackerModule";
  }

  //
  // Tracker

  @ReactMethod
  public void getAppKey(Callback callback) {
    String appKey = getDefaultTracker().getAppKey();
    callback.invoke(appKey);
  }

  @ReactMethod
  public void getVisitorId(Callback callback) {
    String visitorId = getDefaultTracker().getVisitorId();
    callback.invoke(visitorId);
  }

  @ReactMethod
  public void track(String eventName, ReadableMap map) {
    getDefaultTracker().track(eventName, toJSONObject(map));
  }

  @ReactMethod
  public void identify(ReadableMap map) {
    getDefaultTracker().identify(toJSONObject(map));
  }

  @ReactMethod
  public void view(String viewName, String title, ReadableMap map) {
    getDefaultTracker().view(viewName, title, toJSONObject(map));
  }

  @ReactMethod
  public void registerFCMToken(String fcmToken) {
    getDefaultTracker().trackFcmToken(fcmToken);
  }

  @ReactMethod
  public void optIn() {
    getDefaultTracker().optIn();
  }

  @ReactMethod
  public void optOut() {
    getDefaultTracker().optOut();
  }

  @ReactMethod
  public void renewVisitorId() {
    getDefaultTracker().renewVisitorId();
  }

  //
  // TrackerJsUtil

  @ReactMethod
  public void appendUserSyncQueryParameter(String url, Callback callback) {
    String appKey = getDefaultTracker().getAppKey();
    String newURL = TrackerJsUtil.appendUserSyncQueryParameter(getReactApplicationContext(), appKey, url);
    callback.invoke(newURL);
  }

  //
  // InAppMessagingManager

  @ReactMethod
  public void isPresenting(Callback callback) {
    InAppMessagingManager manager = getDefaultTracker().getInAppMessagingManager();
    callback.invoke(manager.isPresenting());
  }

  @ReactMethod
  public void dismiss() {
    getDefaultTracker().getInAppMessagingManager().dismiss();
  }

  @ReactMethod
  public void suppress() {
    getDefaultTracker().getInAppMessagingManager().suppress();
  }

  @ReactMethod
  public void unsuppress() {
    getDefaultTracker().getInAppMessagingManager().unsuppress();
  }


  //
  // Variables

  @ReactMethod
  public void fetch(final Callback callback) {
    String appKey = getDefaultTracker().getAppKey();
    Variables.getInstance(getReactApplicationContext(), appKey).fetch(new Variables.CompletionHandler() {
      @Override
      public void onCompleted(boolean b) {
        callback.invoke(b);
      }
    });
  }

  @ReactMethod
  public void variable(String key, Callback callback) {
    String appKey = getDefaultTracker().getAppKey();
    Variables variables = Variables.getInstance(getReactApplicationContext(), appKey);

    Variable variable = variables.getVariable(key);

    WritableMap map = Arguments.createMap();
    map.putString("key", key);
    map.putString("campaign_id", variable.getCampaignId());
    map.putString("shorten_id", variable.getShortenId());
    map.putString("string", variable.getString(null));

    callback.invoke(map);
  }

  @ReactMethod
  public void trackVariables(ReadableArray vals, String eventName, ReadableMap values) {
    List<Variable> variables = new ArrayList<>();

    for (int i = 0; i < vals.size(); i++) {
      ReadableMap map = vals.getMap(i);
      String campaignId = map.getString("campaign_id");
      String shortenId = map.getString("shorten_id");
      variables.add(new Variable(null, campaignId, shortenId));
    }

    String appKey = getDefaultTracker().getAppKey();
    Variables.getInstance(getReactApplicationContext(), appKey).track(variables, eventName, toJSONObject(values));
  }

  //
  // Variable

  @ReactMethod
  public void stringForKey(String key, String defaultValue, Callback callback) {
    String appKey = getDefaultTracker().getAppKey();
    Variable variable = Variables.getInstance(getReactApplicationContext(), appKey).getVariable(key);

    callback.invoke(variable.getString(defaultValue));
  }

  @ReactMethod
  public void integerForKey(String key, Integer defaultValue, Callback callback) {
    String appKey = getDefaultTracker().getAppKey();
    Variable variable = Variables.getInstance(getReactApplicationContext(), appKey).getVariable(key);

    callback.invoke((int)variable.getLong(defaultValue));
  }

  @ReactMethod
  public void doubleForKey(String key, Double defaultValue, Callback callback) {
    String appKey = getDefaultTracker().getAppKey();
    Variable variable = Variables.getInstance(getReactApplicationContext(), appKey).getVariable(key);

    callback.invoke(variable.getDouble(defaultValue));
  }

  @ReactMethod
  public void boolForKey(String key, Boolean defaultValue, Callback callback) {
    String appKey = getDefaultTracker().getAppKey();
    Variable variable = Variables.getInstance(getReactApplicationContext(), appKey).getVariable(key);

    callback.invoke(variable.getBoolean(defaultValue));
  }

  @ReactMethod
  public void arrayForKey(String key, ReadableArray defaultValue, Callback callback) {
    String appKey = getDefaultTracker().getAppKey();
    Variable variable = Variables.getInstance(getReactApplicationContext(), appKey).getVariable(key);

    JSONArray jsonArray = variable.getJSONArray(toJSONArray(defaultValue));
    if (jsonArray == null) {
      callback.invoke();
    } else {
      try {
        callback.invoke(Converter.convertJsonToArray(jsonArray));
      } catch (JSONException e) {
        callback.invoke();
      }
    }
  }

  @ReactMethod
  public void objectForKey(String key, ReadableMap defaultValue, Callback callback) {
    String appKey = getDefaultTracker().getAppKey();
    Variable variable = Variables.getInstance(getReactApplicationContext(), appKey).getVariable(key);

    JSONObject jsonObject = variable.getJSONObject(toJSONObject(defaultValue));
    if (jsonObject == null) {
      callback.invoke();
    } else {
      try {
        callback.invoke(Converter.convertJsonToMap(jsonObject));
      } catch (JSONException e) {
        callback.invoke();
      }
    }
  }

  private Tracker getDefaultTracker() {
    return Tracker.getInstance(getReactApplicationContext());
  }

  private JSONObject toJSONObject(ReadableMap map) {
    if (map == null) {
      return null;
    }
    return new JSONObject(map.toHashMap());
  }

  private JSONArray toJSONArray(ReadableArray array) {
    if (array == null) {
      return null;
    }
    return new JSONArray(array.toArrayList());
  }
}